python/ql/lib/semmle/python/security/dataflow/LdapInjection.qll

```diff
@@ -8,80 +8,42 @@ import semmle.python.dataflow.new.DataFlow
-    sink = any(LdapExecution ldap).getBaseDn() and
-    (state instanceof Unsafe or state instanceof UnsafeForDn)
-    or
-    sink = any(LdapExecution ldap).getFilter() and
-    (state instanceof Unsafe or state instanceof UnsafeForFilter)
-  }
+    override predicate isSource(DataFlow::Node source) { source instanceof Source }
 
-  override predicate isBarrier(DataFlow::Node node, DataFlow::FlowState state) {
-    // All additional flow steps signify a state change.
-    //    (n, `state`) --> (`node`, s)
-    // Thus, if a node in `state` transitions to `node` and a new state,
-    // then `state` should be blocked at `node`.
-    isAdditionalFlowStep(_, state, node, _)
-  }
+    override predicate isSink(DataFlow::Node sink) { sink instanceof DnSink }
 
-  override predicate isAdditionalFlowStep(
-    DataFlow::Node node1, DataFlow::FlowState state1, DataFlow::Node node2,
-    DataFlow::FlowState state2
-  ) {
-    exists(LdapDnEscaping ldapDnEsc |
-      node1 = ldapDnEsc.getAnInput() and
-      node2 = ldapDnEsc.getOutput()
-    |
-      state1 instanceof Unsafe and
-      state2 instanceof UnsafeForFilter
-      or
-      state1 instanceof UnsafeForDn and
-      state2 instanceof Safe
-    )
-    or
-    exists(LdapFilterEscaping ldapFilterEsc |
-      node1 = ldapFilterEsc.getAnInput() and
-      node2 = ldapFilterEsc.getOutput()
-    |
-      state1 instanceof Unsafe and
-      state2 instanceof UnsafeForDn
-      or
-      state1 instanceof UnsafeForFilter and
-      state2 instanceof Safe
-    )
+    override predicate isSanitizer(DataFlow::Node node) { node instanceof DnSanitizer }
+
+    override predicate isSanitizerGuard(DataFlow::BarrierGuard guard) {
+      guard instanceof DnSanitizerGuard
+    }
   }
-}
 
-/** A flow satte signifying generally unsafe input. */
-class Unsafe extends DataFlow::FlowState {
-  Unsafe() { this = "Unsafe" }
-}
+  class FilterConfiguration extends TaintTracking::Configuration {
+    FilterConfiguration() { this = "LdapFilterInjection" }
 
-/** A flow state signifying input that is only unsafe for DNs. */
-class UnsafeForDn extends DataFlow::FlowState {
-  UnsafeForDn() { this = "UnsafeForDn" }
-}
+    override predicate isSource(DataFlow::Node source) { source instanceof Source }
 
-/** A flow state signifying input that is only unsafe for filter strings. */
-class UnsafeForFilter extends DataFlow::FlowState {
-  UnsafeForFilter() { this = "UnsafeForFilter" }
-}
+    override predicate isSink(DataFlow::Node sink) { sink instanceof FilterSink }
 
-/**
- * A flow state that signifies safe input.
- * Including this makes `isAdditionalFlowStep` and `isBarrier` simpler.
- */
-class Safe extends DataFlow::FlowState {
-  Safe() { this = "Safe" }
+    override predicate isSanitizer(DataFlow::Node node) { node instanceof FilterSanitizer }
+
+    override predicate isSanitizerGuard(DataFlow::BarrierGuard guard) {
+      guard instanceof FilterSanitizerGuard
+    }
+  }
+
+  import DataFlow::PathGraph
```

Comment:
this part should be in the query itself, otherwise you will always expose the extra query predicates by importing this file :disappointed:

Reply:
I am not sure I understand this (that line was also in the query and superfluous here), but it has disappeared anyway with the rewrite proposed below.